### PR TITLE
First time event targeting

### PR DESCRIPTION
**Pull request by msiebert**
Allows feature flags to detect when a users does an action for the first time and changes the variant for the remainder of the session. To accomplish this:

1. The evaluation endpoint sends down "pending first time events" that contain the event and properties (in JSONLogic form to match the runtime property stuff used elsewhere in feature flags)
2. The SDK checks each tracked event. If it matches, we update the cached flag variants with the variant in the pending event so that `getVariant` starts to return the activated value
3. The SDK sends a call to a new endpoint that marks the user for the cohort so they no longer fall into the "pending" state

The following edge cases are also handled when fetching flags:
1. If a flag does not have an assigned variant when first loading flags, we create the flag in the flags map when activating
2. When we refresh flags, we keep the activated value
3. When we refresh flags, if a flag that has been activated is not present in the response, we add it back in

The backend work that the SDK interacts with can be found in [this PR](https://github.com/mixpanel/analytics/pull/85277)

### CODE REVIEW

File: MixpanelDemo/MixpanelDemoTests/MixpanelFeatureFlagTests.swift [hunk 1]
```diff
@@ -1701,4 +1915,286 @@ class FeatureFlagManagerTests: XCTestCase {
...
+      eventProperties: ["plan": "premium"],
+      filters: filters,
+      pendingVariant: pendingVariant,
+      initialVariant: initialVariant,
+      cohortHash: "hash789"
+    ) { mockMgr in
+      let flag = mockMgr.flags?["case-test"]
+      XCTAssertEqual(flag?.key, "matched")
+    }
+  }
```
[the remark this answers is not in the document]
Comment (msiebert) [reply MixpanelDemo/MixpanelDemoTests/MixpanelFeatureFlagTests.swift]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/mixpanel/mixpanel-swift/pull/685#discussion_r2579069481)

**Review by jaredmixpanel — COMMENTED** [review]
# PR #685 Review Summary: First Time Event Targeting

**Changes:** +1087/-180 lines across 7 files

---

## Critical Issues (Must Fix Before Merge)

### 1. **Thread Safety Violation** ⚠️
- **File:** `Sources/FeatureFlags.swift`
- **Issue:** Uses serial DispatchQueue instead of `ReadWriteLock`, breaking project-wide thread safety conventions
- **CLAUDE.md Violation:** "ALWAYS protect shared state" with ReadWriteLock
- **Impact:** Inconsistent with rest of SDK (Track, MixpanelInstance, People all use ReadWriteLock)

### 2. **All Logging Uses `print()` Instead of `MixpanelLogger`** ⚠️
- **File:** `Sources/FeatureFlags.swift` (25+ instances)
- **Issue:** `print()` statements go to stdout but are NOT captured by:
  - MixpanelLogger infrastructure
  - Production crash reporting
  - Any registered logging handlers
- **Impact:** All debugging messages invisible in production

### 3. **Silent JsonLogic Evaluation Failures** ⚠️
- **File:** `Sources/FeatureFlags.swift:835-838`
- **Code:**
  ```swift
  guard let result: Bool = try? applyRule(rulesString, to: dataString),
        result == true else {
      continue  // NO LOGGING HERE
  }
  ```
- **Impact:** Cannot distinguish "rule evaluated to false" from "rule threw an error"

---

## Important Issues (Should Fix)

### 4. **Missing MixpanelType Validation**
- **File:** `Sources/Track.swift:86`
- **Issue:** Properties passed to `checkFirstTimeEvents` not validated against MixpanelType
- **CLAUDE.md Violation:** "Always validate: assertPropertyTypes(properties)"

### 5. **Missing `[weak self]` in Network Completion Handlers**
- **File:** `Sources/FeatureFlags.swift:867-923`
- **Issue:** `recordFirstTimeEvent()` completion handlers capture self strongly
- **Impact:** Potential retain cycles

### 6. **Test Coverage Gaps**
| Gap | Priority | Risk |
|-----|----------|------|
| No integration test for Track.swift → checkFirstTimeEvents flow | High | Regression risk |
| No tests for `recordFirstTimeEvent()` network failures | High | Silent failures |
| No tests for multiple pending events on same flag | Medium | Race conditions |
| `createPendingEvent()` hash inconsistency in tests | Medium | Misleading assertions |

### 7. **API Surface Concern**
- **Issue:** `checkFirstTimeEvents` added to **public** `MixpanelFlags` protocol
- **Recommendation:** This is called automatically from tracking flow - consider making internal

---

## Suggestions (Nice to Have)

### 8. **Type Design Improvements**
| Improvement | Benefit |
|-------------|---------|
| Add `Hashable` to `PendingFirstTimeEvent` with `compositeKey` | Cleaner identity semantics |
| Use `Properties` typealias instead of `[String: Any]` | Type consistency |
| Make internal state `private` with test accessors | Stronger encapsulation |

### 9. **Error Message Improvements**
- Separate compound guard failures into individual checks with specific messages
- Include flag names, event keys in error context

---

## Strengths ✅

1. **Well-structured feature design** - Clean separation of parsing, matching, activation, and backend notification
2. **Comprehensive test coverage for happy paths** - Response parsing, event matching, filter evaluation, activation state
3. **Edge case handling** - Preserves activated variants on refresh, keeps orphaned flags
4. **Follows existing patterns** - Uses `accessQueue` for serialization, `AnyCodable` for dynamic types
5. **Good test helpers** - Extracted `setupReadyFlags()`, `expectTracking()`, `verifyTrackedEvent()` reduce boilerplate

---

## Recommended Action

**Fix critical issues before merge:**

1. Replace all `print()` calls with `MixpanelLogger.debug/info/warn/error`
2. Add proper error handling for JsonLogic evaluation (use `do-catch` instead of `try?`)
3. Consider thread safety approach - either document why serial queue is acceptable here, or migrate to ReadWriteLock

**Address before merge or in follow-up:**

4. Add integration test for Track.swift flow
5. Add `[weak self]` to network completion handlers
6. Add test for `recordFirstTimeEvent` network failure

---

## Files Changed Summary

| File | Changes | Status |
|------|---------|--------|
| `Sources/FeatureFlags.swift` | +314 lines (main implementation) | Needs fixes |
| `Sources/Track.swift` | +5 lines (integration hook) | OK |
| `MixpanelFeatureFlagTests.swift` | +707 lines (tests) | Needs minor fixes |
| `Package.swift` | +9 lines (jsonlogic dep) | OK |
| `Mixpanel-swift.podspec` | +1 line (jsonlogic dep) | OK |
| `project.pbxproj` | +49 lines (Xcode config) | OK |
| `MixpanelDemoTests.swift` | +2 lines (protocol stub) | OK |

---

*Review generated with assistance from automated code analysis.*

**Comment by msiebert** [comment]
### Code review

Found 1 issue:

1. Race condition allows duplicate event activations - In `checkFirstTimeEvents`, the activated events check uses a stale copy (line 823) read outside the write lock. Two concurrent calls with the same event can both pass the "already activated" check using their stale copies, then both acquire the write lock sequentially and activate the same event multiple times. This causes duplicate `_trackFlagIfNeeded` calls and duplicate `recordFirstTimeEvent` backend calls, violating the "first-time" semantics. The original serial queue implementation prevented this by serializing all checks and writes.

https://github.com/mixpanel/mixpanel-swift/blob/d08d10850c66c1cfe33d4cb5456ff19aaceb26ef/Sources/FeatureFlags.swift#L815-L878

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: Package.swift [hunk 2]
```diff
@@ -13,9 +13,18 @@ let package = Package(
   products: [
     .library(name: "Mixpanel", targets: ["Mixpanel"])
   ],
+  dependencies: [
+    .package(
+      url: "https://github.com/advantagefse/json-logic-swift",
+      from: "1.2.0"
+    )
```
[the remark this answers is not in the document]
Comment (msiebert) [reply Package.swift]: @copilot please add this documentation

File: Package.swift [hunk 3]
```diff
@@ -13,9 +13,18 @@ let package = Package(
   products: [
     .library(name: "Mixpanel", targets: ["Mixpanel"])
   ],
+  dependencies: [
+    .package(
+      url: "https://github.com/advantagefse/json-logic-swift",
```
[the remark this answers is not in the document]
Comment (tylerjroach) [reply Package.swift]: This has already been discussed and not a blocker for now. While it would be nice to have, this gives us what we need immediately.
